src/routes/api/embed/project/[projectUrl]/support.png/+server.ts

```diff
@@ -0,0 +1,106 @@
+import type { RequestHandler, RouteParams } from './$types';
+import { error } from '@sveltejs/kit';
+import puppeteer from 'puppeteer';
+import { redis } from '$lib/../routes/api/redis';
+import {
+  getSupportButtonOptions,
+  SupportButtonStat,
+  SupportButtonText,
+  type SupportButtonOptions,
+} from '$lib/components/project-support-button/project-support-button';
+
+const REPLACE_PNG_REGEX = /(\.png\/?)(\?.*|$)/;
+const CACHE_KEY_PREFIX = 'support-button';
+
+const getCacheExpiration = (options: SupportButtonOptions): number => {
+  // Cache dynamic data for 6 hours
+  if (options.stat !== SupportButtonStat.none) {
+    return 60 * 60 * 6;
+  }
+
+  return Infinity;
+};
+
+const getCacheKey = (options: SupportButtonOptions, params: RouteParams): string => {
+  // buttons without project details and $ support or dependencies can be cached
+  // for all projects to use.
+  let key = `${CACHE_KEY_PREFIX}-${options.style}-${options.background}-${options.text}`;
+  if (options.text === SupportButtonText.project) {
+    key += `-${encodeURIComponent(params.projectUrl)}`;
+  }
+
+  // AND the button will only be cached for 6 hours
+  if (options.stat !== SupportButtonStat.none) {
+    key += `-${options.stat}`;
+  }
+
+  return key;
+};
+
+export const GET: RequestHandler = async ({ url, params }) => {
+  let browser;
+  try {
+    // drips.network/embed/project.png/support.png/?background=dark
+    // ==> drips.network/embed/project.png/support?background=dark
+    const imageUrl = url.href.replace(REPLACE_PNG_REGEX, '$2');
+    // the URL should always be re-written
+    if (imageUrl === url.href) {
+      return error(400);
+    }
+
+    const buttonOptions = getSupportButtonOptions(url);
+    // Try to fetch the pre-rendered image from cache
+    const cacheKey = getCacheKey(buttonOptions, params);
+    const cachedImageBase64 = redis && (await redis.get(cacheKey));
+    if (cachedImageBase64) {
+      const cachedImageBuffer = Buffer.from(cachedImageBase64, 'base64');
+      return new Response(cachedImageBuffer, {
+        status: 200,
+        headers: new Headers({ 'Content-Type': 'image/png' }),
+      });
+    }
+
+    browser = await puppeteer.launch({
+      // Dockerfile deployment requires different executablePath
+      ...(process.env.NODE_ENV === 'production' && {
+        executablePath: '/usr/bin/google-chrome-stable',
+      }),
+      headless: true,
+      args: ['--no-sandbox', '--disable-setuid-sandbox'],
+    });
```

Comment:
I thought about this! I'm just closing it in the name of cleaning up resources really. I figured caching would reduce the hit launching the browser would cause anyways? I could imagine that just closing the page and keeping the browser instance around would be speedy. I also had some reservations about how the deployment environment would treat a long-running browser instance. If it were to be killed, we would need to handle that in some way, blah blah

Reply:
yep, all valid points

let's revisit if / when we see performance being a problem